### PR TITLE
feat(react-color-picker): added context to the ColorPicker

### PR DESCRIPTION
**Pull request by ValentinaKozlova**
image

### CODE REVIEW

File: packages/react-components/react-color-picker-preview/library/src/components/ColorPicker/useColorPicker.ts [hunk 1]
```diff
@@ -12,20 +12,29 @@ import type { ColorPickerProps, ColorPickerState } from './ColorPicker.types';
  * @param ref - reference to root HTMLDivElement of ColorPicker
  */
 export const useColorPicker_unstable = (props: ColorPickerProps, ref: React.Ref<HTMLDivElement>): ColorPickerState => {
+  const { color, onColorChange, ...rest } = props;
+
+  const requestChange: ColorPickerState['requestChange'] = useEventCallback((event, data) => {
+    onColorChange?.(event, {
```
Comment (ValentinaKozlova) [inline packages/react-components/react-color-picker-preview/library/src/components/ColorPicker/useColorPicker.ts:18]: In following PRs color manipulation will be here

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts [hunk 2]
```diff
@@ -16,7 +17,8 @@ export const useColorSliderState_unstable = (state: ColorSliderState, props: Col
   'use no memo';
 
   const { dir } = useFluent();
-  const { defaultValue, min = 0, max = MAX_COLOR_HUE, onChange, value } = props;
+  const { channel = 'hue', defaultValue, min = 0, max = MAX_COLOR_HUE, onChange, value } = props;
+  const ctxOnChange = useColorPickerContextValue_unstable(ctx => ctx.requestChange);
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts:21]:
```suggestion
  const onChangeFromContext = useColorPickerContextValue_unstable(ctx => ctx.requestChange);
  const { channel = 'hue', defaultValue, min = 0, max = MAX_COLOR_HUE, onChange = onChangeFromContext, value } = props;
```

nit

File: packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts [hunk 3]
```diff
@@ -26,16 +28,17 @@ export const useColorSliderState_unstable = (state: ColorSliderState, props: Col
   const clampedValue = clamp(currentValue, min, max);
   const valuePercent = getPercent(clampedValue, min, max);
 
+  const propsOnChange = onChange ?? ctxOnChange;
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts:31]:
```suggestion
```

File: packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts [hunk 4]
```diff
@@ -26,16 +28,17 @@ export const useColorSliderState_unstable = (state: ColorSliderState, props: Col
...
     const newValue = Number(ev.target.value);
     setCurrentValue(clamp(newValue, min, max));
 
-    if (inputOnChange && inputOnChange !== (onChange as unknown as React.ChangeEventHandler<HTMLInputElement>)) {
+    if (inputOnChange && inputOnChange !== (propsOnChange as unknown as React.ChangeEventHandler<HTMLInputElement>)) {
       inputOnChange(ev);
-    } else if (onChange) {
-      onChange(ev, { type: 'change', event: ev, value: newValue });
+    } else if (propsOnChange) {
+      propsOnChange(ev, { type: 'change', event: ev, value: newValue, channel });
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts:41]: Why this is needed? Why it cannot be following?

```
inputOnChange(ev);
propsOnChange(ev, { type: 'change', event: ev, value: newValue, channel });
```

File: packages/react-components/react-color-picker-preview/library/etc/react-color-picker-preview.api.md [hunk 5]
```diff
@@ -53,16 +53,22 @@ export const ColorPicker: ForwardRefComponent<ColorPickerProps>;
 // @public (undocumented)
 export const colorPickerClassNames: SlotClassNames<ColorPickerSlots>;
 
+// @public (undocumented)
+export type ColorPickerOnSelectEventHandler = EventHandler<ColorPickerOnChangeData>;
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/etc/react-color-picker-preview.api.md:57]: Let's avoid exposing such types, it's expected to use `ColorPickerProps['onColorChange']` for example

File: packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts (hunk 3 above)
Comment (ValentinaKozlova) [reply packages/react-components/react-color-picker-preview/library/src/components/ColorSlider/useColorSliderState.ts]: fixed

File: packages/react-components/react-color-picker-preview/library/src/components/AlphaSlider/useAlphaSliderState.ts [hunk 6]
```diff
@@ -26,16 +28,21 @@ export const useAlphaSliderState_unstable = (state: AlphaSliderState, props: Alp
...
-      onChange(ev, { type: 'change', event: ev, value: newValue });
-    }
-  });
+  const _onChange: React.ChangeEventHandler<HTMLInputElement> = useEventCallback(
+    mergeCallbacks(
+      event => {
+        const newValue = Number(event.target.value);
+        setCurrentValue(clamp(newValue, min, max));
+        inputOnChange && inputOnChange(event);
+        onChange && onChange(event, { type: 'change', event, value: newValue, channel });
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/src/components/AlphaSlider/useAlphaSliderState.ts:37]:
```suggestion
        inputOnChange?.(event);
        onChange?.(event, { type: 'change', event, value: newValue, channel });
```

File: packages/react-components/react-color-picker-preview/library/src/components/AlphaSlider/useAlphaSliderState.ts [hunk 7]
```diff
@@ -26,16 +28,21 @@ export const useAlphaSliderState_unstable = (state: AlphaSliderState, props: Alp
...
-    setCurrentValue(clamp(newValue, min, max));
-
-    if (inputOnChange && inputOnChange !== (onChange as unknown as React.ChangeEventHandler<HTMLInputElement>)) {
-      inputOnChange(ev);
-    } else if (onChange) {
-      onChange(ev, { type: 'change', event: ev, value: newValue });
-    }
-  });
+  const _onChange: React.ChangeEventHandler<HTMLInputElement> = useEventCallback(
+    mergeCallbacks(
```
Comment (layershifter) [inline packages/react-components/react-color-picker-preview/library/src/components/AlphaSlider/useAlphaSliderState.ts:32]: `mergeCallbacks()` seems to be redundant there, you can do:

```
      event => {
        const newValue = Number(event.target.value);
        setCurrentValue(clamp(newValue, min, max));
        inputOnChange && inputOnChange(event);
        onChange && onChange(event, { type: 'change', event, value: newValue, channel });
        ctxOnChange(event, {
          value: Number(event.target.value),
          channel,
        }),
```

Comment (ValentinaKozlova) [reply packages/react-components/react-color-picker-preview/library/src/components/AlphaSlider/useAlphaSliderState.ts]: fixed

**Review by layershifter — APPROVED** [review]

**Review by mainframev — APPROVED** [review]

**Pull request merged**